### PR TITLE
Allow multi key remappings in config file

### PR DESCRIPTION
**Pull request by sudormrfbin**
Allows configuration like this:

```toml
[keys.normal.g.l]
"d" = "goto_definition"  # gld
"r" = "goto_reference"  # glr

[keys.normal.z]
"$" = "goto_file_end"  # z$

[keys.insert.j]
"k" = "normal_mode"  # jk
```

Or alternatively in inline table syntax:

```toml
[keys.normal]
# gld -> definitions, glr -> references
"g" = {"l" = {"d" = "goto_definition", "r" = "goto_reference"} }
"z" = {"$" = "goto_file_end"}

[keys.insert]
"j" = {"k" = "normal_mode"}
```

We can't use something like `"jk" = "normal_mode"` since that will collide with key names like `"backspace"` and `"esc"`.

Key mappings are represented in a tree like structure, which can be useful in the future to dynamically show infobox contents (currently the keys are hardcoded, and user mappings cannot be shown).

### CODE REVIEW

**Comment by kirawi** [comment]
So I presume you would do this for 3 keys?:
`"e" = { "s" = { "c" = "normal_mode" } }`

**Comment by sudormrfbin** [comment]
> So I presume you would do this for 3 keys?:
> `"e" = { "s" = { "c" = "normal_mode" } }`

@kirawi Exactly, it can be arbitrarily nested to any level. An alternative syntax (supported by the toml spec):

```toml
[keys.insert.e.s]
"c" = "normal_mode"
```

**Comment by pickfire** [comment]
```toml
[keys.insert.e.s]
"c" = "normal_mode"
```
This does seemed better, is it possible to switch to this?

**Comment by sudormrfbin** [comment]
Both methods are supported by the parser, it's in the defined toml spec :)

File: helix-term/src/ui/editor.rs [hunk 1]
```diff
@@ -559,19 +563,72 @@ impl EditorView {
...
+            match keynode {
+                KeyNode::KeyCommand(command) => {
+                    command.execute(cxt);
+                    self.current_subkeymap = None;
+                    self.subkeymap_level = 0;
+                }
+                KeyNode::SubKeymap(subkeymap) => {
+                    self.current_subkeymap = Some(subkeymap.clone());
+                    self.subkeymap_level += 1;
+                }
```
Comment (pickfire) [inline helix-term/src/ui/editor.rs:592]: I don't think this is a good way to solve this. Ideally we want to be able to show infobox as well since it requires similar thing.

**Review by pickfire — CHANGES_REQUESTED** [review]
I think it is better to integrate this with infobox to kill two birds with one stone.

- able to change infobox keys
- able to show infobox for user custom keys

Maybe you can look at how infobox is implemented, I have this in mind so I parse the keys the same way the config is currently being parsed.

https://github.com/helix-editor/helix/blob/e2bcef718abfe9df8b3169e6017053506de5e1d1/helix-term/src/commands.rs#L3691-L3714

Do you want to give it a try? If anything can ping me.

File: helix-term/src/ui/editor.rs (hunk 1 above)
Comment (sudormrfbin) [reply helix-term/src/ui/editor.rs:592]: Since we can now specify multi key mappings, I was thinking of moving all of the builtin ones like `gy`, `gr`, `<space>b`, etc from the `mode_info` macros to `keymap.rs` and then render the infobox from here (in the `KeyNode::SubKeymap(subkeymap) =>` branch) instead of doing it from the macro generated mode functions. I think it'll solve the problem since the default keys shown in infobox after this change aren't a special case and are essentially defined the same way as user defined multi key remappings.

**Comment by sudormrfbin** [comment]
The infobox was one of the reasons I went with a tree to represent the keys since it would be easy to get the next set of possible keys after a keypress; this is how I was thinking of doing it:

```diff
--- a/helix-term/src/keymap.rs
+++ b/helix-term/src/keymap.rs
@@ -195,7 +195,18 @@ impl Default for Keymaps {
             ctrl!('o') => KeyCommand(Command::jump_backward),
             // ctrl!('s') KeyCommand(ommand)::save_selection,
 
-            key!(' ') => KeyCommand(Command::space_mode),
+            key!(' ') => SubKeymap(hashmap!(
+                key!('f') => KeyCommand(Command::file_picker),
+                key!('b') => KeyCommand(Command::buffer_picker),
+                key!('s') => KeyCommand(Command::symbol_picker),
+                key!('w') => KeyCommand(Command::window_mode),
+                key!('y') => KeyCommand(Command::yank_joined_to_clipboard),
+                key!('Y') => KeyCommand(Command::yank_main_selection_to_clipboard),
+                key!('p') => KeyCommand(Command::paste_clipboard_after),
+                key!('P') => KeyCommand(Command::paste_clipboard_before),
+                key!('R') => KeyCommand(Command::replace_selections_with_clipboard),
+                key!(' ') => KeyCommand(Command::keep_primary_selection)
+            )),
```

So instead of using the `mode_info` macro (which we would have to treat as a special case since it does it's own key handling waiting for the next key), all the multi key mappings are defined in `keymap.rs`. The descriptions can be stored as a field the `Command` struct along with the name and callback function. I'll try playing around with it :)

**Comment by pickfire** [comment]
I don't think tree is the issue, I think both specifying the message during compile-time works but I think we should reuse the `mode_info!` or have something similar to replace the current mechanism, that is able to make it more ergonomics from our side.

But there is a benefit to let user specify their message, because we don't know how the user structure their keys so it is better to let users override the description for the infobox as well, which is why I think the description should be in runtime config, because we don't know how user want to structure the stuff, and the infobox should depends on how it is structured to have custom text.

Although infobox integration is not necessary now but at the very least I would like to see the replacement of `key!()` to use our own parser, so that we can specify `"space"` and the rest like the infobox macro.

**Comment by archseer** [comment]
I think this is a good feature, but it shouldn't be bolted onto the current data structures.

We should implement a custom trie structure for keys, and then fetching the value either returns a final node (a command) or a node with more branches, which can then be used to render the autoinfo in a generic way for any subcommand.

**Comment by archseer** [comment]
Example from iota https://github.com/gchp/iota/blob/8f35dd5cba2f07077590619760e62dbb74a71fab/src/iota/keymap.rs

**Comment by sudormrfbin** [comment]
@pickfire I think we can start with non-configurable infobox descriptions since the user cannot make custom commands yet, and if the default descriptions need changing they should probably be changed in the repo itself. I'll look into replacing `key!` (and the whole keymap definition method) with another macro which doesn't look so crowded.

@archseer You mean storing the state within `Keymaps` like in iota and removing it from `EditorView` ? That does seem cleaner, also a proper trie will represent it better :+1:

**Comment by archseer** [comment]
That said, I took a look and the trie in iota looks a lot like what we have now (a bunch of hashmaps) :) so we can probably just tweak the current implementation.

Yeah I think the way to go is improve the `Keymap` type so that we can define the mode mappings there without the need for custom commands. You can still store the state (keypresses up to now) inside the EditorView, basically you push the key onto the `keys` vec, traverse and see what you get. If it's a command you can clear the `keys` and execute, if it's a branch you keep the keys and render an infobox instead.

In order for the descriptions to work we'll have to define them per command (instead of within the `mode!` macro) but that can be done similar to how we do it for `:commands`.

**Comment by cessen** [comment]
This PR is heading in a direction that I'm really excited about, but I think can potentially go even further.  I'm not sure if we want to tackle this in this PR, but here's what I was thinking:

The keymap tree wouldn't just handle sequential key presses, but also modes.  Basically, if you think of normal mode as being the "root", then the keymap tree branches off from there, and that tree can have both:

- Transient nodes: used for sequential key combos, like this PR is currently enabling.  These nodes can optionally include names for displaying in the info box pop-up.
- Sticky/mode nodes: used for defining custom editor modes.  When you press a key corresponding to a sticky node, you're there permanently until you explicitly escape out of it.  These nodes would include a name for displaying in the status line (e.g. `SEL`), as well as a key that can be used in themes.

This would allow nearly complete customization of Helix's interaction model.

The main question I'm still unsure of with the above idea is how insert mode fits into all of this.  Possible each sticky node has a boolean flag that specifies whether it's an "insert-like" mode or a "normal-like" mode.  Not sure.

(Sorry if this is off-topic for the PR... I'm just really excited about getting the Helix interaction model super customizable, ha ha.)

File: helix-term/src/keymap.rs [hunk 2]
```diff
@@ -62,188 +153,187 @@ impl DerefMut for Keymaps {
 
 impl Default for Keymaps {
     fn default() -> Keymaps {
+        use KeyTrie::*;
         let normal = hashmap!(
```
Comment (pickfire) [inline helix-term/src/keymap.rs:157]: Can we change hashmap macro to automatically add the `Leaf`?

**Review by pickfire — CHANGES_REQUESTED** [review]
I think we can just defer infobox integration but I would like to see `hashmap!` macro being updated to add the `Leaf`.

File: helix-term/src/keymap.rs (hunk 2 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs]: I made a new `keymap!` macro inspired by `mode_info!` which takes it a bit further :) https://github.com/helix-editor/helix/pull/454/files#diff-8b8f04b030b7fdea89f3d9ec2b7aeb6a33959ee4c6cb27f0d574f115b474d151R157-R161

File: helix-term/src/keymap.rs [hunk 3]
```diff
@@ -62,247 +154,300 @@ impl DerefMut for Keymaps {
...
+            "w" => move_next_word_start,
+            "b" => move_prev_word_start,
+            "e" => move_next_word_end,
+
+            "W" => move_next_long_word_start,
+            "B" => move_prev_long_word_start,
+            "E" => move_next_long_word_end,
+
+            "v" => select_mode,
+            "g" => {
```
Comment (sudormrfbin) [inline helix-term/src/keymap.rs:187]: Goto line number doesn't work now since there is no separate function for goto mode. Adding a function for goto mode only will require treating it as a special case especially for handling the infobox and user remappings. A solution is to add it to `gg` (keybind would change from `82g` to `82gg`, like in vim).

**Comment by sudormrfbin** [comment]
> * Sticky/mode nodes: used for defining custom editor modes.  When you press a key corresponding to a sticky node, you're there permanently until you explicitly escape out of it.  These nodes would include a name for displaying in the status line (e.g. `SEL`), as well as a key that can be used in themes.

@cessen That's a neat idea; I suppose it's similar to emacs major modes (I don't use emacs so not sure) ?

> I think we can just defer infobox integration but I would like to see `hashmap!` macro being updated to add the `Leaf`.

@pickfire I moved the keys from `mode_info!` definitions into `keymap!` so the keys work but infobox doesn't; I'll add the infobox integration soon.

> Yeah I think the way to go is improve the `Keymap` type so that we can define the mode mappings there without the need for custom commands. You can still store the state (keypresses up to now) inside the EditorView, basically you push the key onto the `keys` vec, traverse and see what you get. If it's a command you can clear the `keys` and execute, if it's a branch you keep the keys and render an infobox instead.

I made `Keymap` itself stateful instead of deferring to `EditorView`, and infobox is todo.

File: helix-term/src/keymap.rs (hunk 3 above)
Comment (pickfire) [reply helix-term/src/keymap.rs:187]: Can't we add hooks?

File: helix-term/src/keymap.rs [hunk 4]
```diff
@@ -62,247 +154,300 @@ impl DerefMut for Keymaps {
...
+                    "v" => vsplit,
+                    "q" => wclose
+                },
+                "y" => yank_joined_to_clipboard,
+                "Y" => yank_main_selection_to_clipboard,
+                "p" => paste_clipboard_after,
+                "P" => paste_clipboard_before,
+                "R" => replace_selections_with_clipboard,
+                "space" => keep_primary_selection
+            },
```
Comment (pickfire) [inline helix-term/src/keymap.rs:323]: Err, doesn't this means the infobox and generated doc will be gone?

File: helix-term/src/keymap.rs (hunk 3 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs:187]: But that would make it impossible to have any other key under `g` to take a count, and the user might remap some function to it that takes a count.

**Comment by cessen** [comment]
> @cessen That's a neat idea; I suppose it's similar to emacs major modes (I don't use emacs so not sure) ?

I don't use emacs either, so no idea, ha ha.  I just have a hope for modes themselves to also be fully customizable in Helix, and that's one idea for how to accomplish that in a relatively coherent way.  But maybe another approach would be better.

File: helix-term/src/keymap.rs (hunk 3 above)
Comment (pickfire) [reply helix-term/src/keymap.rs:187]: The keys under g from what I see all does not need a count.

File: helix-term/src/keymap.rs (hunk 4 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs]: Added infobox back in in d4ed5e31f201c4c17d54c0353b0bd9cf236182a7, headings need to be worked in, will add it soon.

File: helix-term/src/keymap.rs (hunk 3 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs:197]: The default ones don't, I meant the case as a user remapping some other function that takes count to a key under `g`. I will probably do this myself with `ge` to be vim-like (go to end of previous word) and it would take a count.

File: helix-term/src/commands.rs [hunk 5]
```diff
@@ -138,13 +125,21 @@ fn align_view(doc: &Document, view: &mut View, align: Align) {
...
-pub struct Command(&'static str, fn(cx: &mut Context));
+pub struct Command {
+    name: &'static str,
+    fun: fn(cx: &mut Context),
+    doc: &'static str,
+}
 
 macro_rules! commands {
-    ( $($name:ident),* ) => {
+    ( $($name:ident, $doc:literal),* ) => {
```
Comment (pickfire) [inline helix-term/src/commands.rs:135]: Can you please make this take in a `#[doc($doc:literal)]` like how it was previously? Doing `///` might be better. Later I have an idea to extend that to large description without changing anything.

File: helix-term/src/ui/editor.rs [hunk 6]
```diff
@@ -569,9 +572,22 @@ impl EditorView {
         cxt: &mut commands::Context,
         event: KeyEvent,
     ) -> Option<KeymapResult> {
+        self.autoinfo = None;
         match self.keymaps.get_mut(&mode).unwrap().get(event) {
             KeymapResult::Matched(command) => command.execute(cxt),
-            KeymapResult::Pending(_map) => { /* render infobox */ }
+            KeymapResult::Pending(map) => {
+                let mut body = BTreeMap::new();
```
Comment (pickfire) [inline helix-term/src/ui/editor.rs:579]: Why `BTreeMap` instead of `HashMap`?

File: helix-term/src/ui/editor.rs [hunk 7]
```diff
@@ -569,9 +572,22 @@ impl EditorView {
...
+                for (key, trie) in map.into_iter() {
+                    let desc = match trie {
+                        KeyTrie::Leaf(cmd) => cmd.doc(),
+                        KeyTrie::Node(_) => "",
```
Comment (pickfire) [inline helix-term/src/ui/editor.rs:583]: We also need to have doc for node, it will be shown as the title.

File: helix-term/src/ui/editor.rs [hunk 8]
```diff
@@ -569,9 +572,22 @@ impl EditorView {
...
+                    };
+                    // FIXME: multiple keys are ordered randomly (use BTreeSet)
```
Comment (pickfire) [inline helix-term/src/ui/editor.rs:585]: The `keymaps` should preserve the order, it uses `HashMap` which does not.

File: helix-term/src/ui/editor.rs [hunk 9]
```diff
@@ -788,9 +804,8 @@ impl Component for EditorView {
             );
         }
 
-        if let Some(info) = std::mem::take(&mut cx.editor.autoinfo) {
+        if let Some(ref info) = self.autoinfo {
             info.render(area, surface, cx);
-            cx.editor.autoinfo = Some(info);
```
Comment (pickfire) [inline helix-term/src/ui/editor.rs:757]: Nice, I didn't do this last time.

File: helix-view/src/editor.rs [hunk 10]
```diff
@@ -66,7 +64,6 @@ impl Editor {
             theme_loader: themes,
             registers: Registers::default(),
             clipboard_provider: get_clipboard_provider(),
-            autoinfo: None,
```
Comment (pickfire) [inline helix-view/src/editor.rs:69]: If you removed these how are you planning to show autoinfo?

File: helix-term/src/config.rs [hunk 11]
```diff
@@ -42,22 +43,22 @@ fn parsing_keymaps_config_file() {
         toml::from_str::<Config>(sample_keymaps).unwrap(),
         Config {
             keys: Keymaps(hashmap! {
-                Mode::Insert => hashmap! {
+                Mode::Insert => Keymap::new(hashmap! {
```
Comment (pickfire) [inline helix-term/src/config.rs:46]: I am thinking of removing the inner `hashmap!`

File: helix-term/src/ui/editor.rs (hunk 6 above)
Comment (sudormrfbin) [reply helix-term/src/ui/editor.rs]: Hashmaps don't have an inherent order, so every time the infobox is shown it will have the items ordered differently.

File: helix-term/src/ui/editor.rs (hunk 7 above)
Comment (sudormrfbin) [reply helix-term/src/ui/editor.rs]: Fixed 👍🏾

File: helix-view/src/editor.rs (hunk 10 above)
Comment (sudormrfbin) [reply helix-view/src/editor.rs:69]: I moved it into `EditorView` and set it from there when handling key input.

File: helix-term/src/config.rs (hunk 11 above)
Comment (sudormrfbin) [reply helix-term/src/config.rs]: Yeah that test was broken, fixed it

File: helix-term/src/ui/editor.rs (hunk 8 above)
Comment (sudormrfbin) [reply helix-term/src/ui/editor.rs]: The items are ordered correctly now, I meant when the same function is bound to multiple keys (like `zc` and `zz`). It will be `c, z` or `z, c` but consistent in a  single editor session (changes when editor is closed and opened).

File: helix-term/src/commands.rs (hunk 5 above)
Comment (sudormrfbin) [reply helix-term/src/commands.rs:135]: Large description to show in the infobox ? Shouldn't it be kept as compact as possible ?

**Comment by sudormrfbin** [comment]
Misc things I noticed when writing descriptions for commands:

- `Ctrl-X` to invoke completions is not documented on the website
- No default keys assigned for signature help and expand selection (based on tree sitter node)
- No mention of registers in website docs

File: helix-term/src/ui/info.rs [hunk 12]
```diff
@@ -8,7 +8,7 @@ impl Component for Info {
     fn render(&self, viewport: Rect, surface: &mut Surface, cx: &mut Context) {
         let style = cx.editor.theme.get("ui.popup");
         let block = Block::default()
-            .title(self.title)
+            .title(self.title.clone())
```
Comment (archseer) [inline helix-term/src/ui/info.rs:11]: I think you can just pass by ref: `&self.title`

File: helix-view/src/info.rs [hunk 13]
```diff
@@ -1,12 +1,12 @@
 use crate::input::KeyEvent;
 use helix_core::unicode::width::UnicodeWidthStr;
-use std::fmt::Write;
+use std::{collections::BTreeMap, fmt::Write};
 
 #[derive(Debug)]
 /// Info box used in editor. Rendering logic will be in other crate.
 pub struct Info {
     /// Title kept as static str for now.
```
Comment (archseer) [inline helix-view/src/info.rs:8]: Note, this comment is no longer true

**Comment by archseer** [comment]
Great work! I think this is looking much nicer than what we have on `master`, especially the `keymap!` macro. 


>  * No default keys assigned for signature help and expand selection (based on tree sitter node)

Yeah signature help is incomplete and will likely be auto triggered by a hook when a trigger character is used (defined by lsp, usually `(` then retriggered on `,` when typing out arguments.

Expand selection used to be `[` but was replaced. I think we want `[o` and `]o` for contract & expand.

File: helix-view/src/info.rs (hunk 13 above)
Comment (sudormrfbin) [reply helix-view/src/info.rs:8]: Fixed 👍🏾

**Comment by archseer** [comment]
Tested it out on local, it works! Just a few things:

- We should use a BTreeMap to preserve the order of commands in the popup. We grouped them logically before so that goto start of file and end of file sit next to each other on the list.
- The " mode" suffix on each header isn't needed, I'd just use "Goto"/"Window" etc
- Each line in the goto popup now starts with goto. I was thinking an easy hack would be: do all the key entries in the mode start with the mode name? If so, remove the prefix. That way it will show as "implementation" in the goto mode, but "Goto implementation" if it's in a user defined mode.

File: helix-term/src/commands.rs (hunk 5 above)
Comment (pickfire) [reply helix-term/src/commands.rs:135]: Large description won't be shown in infobox. But you can just put a short description like I did for now. If later we want to extend it with large description I have an idea, like how I did with the infobox title.

File: helix-term/src/keymap.rs [hunk 14]
```diff
@@ -62,252 +283,298 @@ impl DerefMut for Keymaps {
...
+            "w" => move_next_word_start,
+            "b" => move_prev_word_start,
+            "e" => move_next_word_end,
+
+            "W" => move_next_long_word_start,
+            "B" => move_prev_long_word_start,
+            "E" => move_next_long_word_end,
+
+            "v" => select_mode,
+            "g" => { "Goto mode"
```
Comment (pickfire) [inline helix-term/src/keymap.rs:315]: Can you please reuse the old titles that you removed rather than creating new ones? Like the following.
```suggestion
            /// goto
            ///
            /// When specified with a count, it will go to that line without entering the mode.
            "g" => {
```
The extra description will only be shown in the rustdoc but the "goto" word will be shown in both rustdoc and infobox.

File: helix-term/src/keymap.rs [hunk 15]
```diff
@@ -62,252 +283,298 @@ impl DerefMut for Keymaps {
...
-            key!('g') => Command::goto_mode,
-            key!(':') => Command::command_mode,
-
-            key!('i') => Command::insert_mode,
-            key!('I') => Command::prepend_to_line,
-            key!('a') => Command::append_mode,
-            key!('A') => Command::append_to_line,
-            key!('o') => Command::open_below,
-            key!('O') => Command::open_above,
+        let normal = keymap!({ "Normal mode"
```
Comment (pickfire) [inline helix-term/src/keymap.rs:286]: Can keymap be `keymap! {`?

File: helix-term/src/keymap.rs [hunk 16]
```diff
@@ -62,252 +283,298 @@ impl DerefMut for Keymaps {
...
+            "left" => move_char_left,
+            "down" => move_line_down,
+            "up" => move_line_up,
+            "right" => move_char_right,
+            "pageup" => page_up,
+            "pagedown" => page_down,
+            "home" => goto_line_start,
+            "end" => goto_line_end_newline,
+
+            "C-x" => completion
```
Comment (pickfire) [inline helix-term/src/keymap.rs:502]: Can you please add additional `,` for the macro?

**Review by pickfire — CHANGES_REQUESTED** [review]
I think the docs and macro can be improved. `"m" => { "Match mode"` looks weird. But at least for now the docs should take from the old one for those that was writtten, I tuned them specifically for the infobox.

File: helix-term/src/keymap.rs (hunk 15 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs:286]: No, since the argument has to be a token tree so that it can recurse properly and detect sub nodes.

File: helix-term/src/keymap.rs (hunk 16 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs:502]: Done 👍🏾

File: helix-term/src/keymap.rs (hunk 14 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs:315]: Reusing the old titles seem better since they are shorter 👍🏾. But specifying it as a doc comment and then having an additional description would be hard to parse since the macro is recursive (also writing descriptions for rustdoc seem redundant since they are not user facing and only developers will see it)

**Comment by sudormrfbin** [comment]
> * We should use a BTreeMap to preserve the order of commands in the popup. We grouped them logically before so that goto start of file and end of file sit next to each other on the list.

@archseer BTreeMap sorts itself by keys, so it won't remember insertion order. I think we can keep a private `Vec` that tracks the order the keys inserted and then use it when constructing the infobox.

> * Each line in the goto popup now starts with goto. I was thinking an easy hack would be: do all the key entries in the mode start with the mode name? If so, remove the prefix. That way it will show as "implementation" in the goto mode, but "Goto implementation" if it's in a user defined mode. I'm not too opinionated here, we might skip this. It's nice because it keeps the popup small though.

I'll try a hand at it :)

> I think the docs and macro can be improved. `"m" => { "Match mode"` looks weird.

My initial preference for the syntax was `"m" => "Match mode" {` and then doc comments, but the recursive nature of the macro (for example everything after `=>` has to be a catch all metavar like `tt` so as to match both leaves and nodes) makes it hard to write a readable macro using those syntaxes.

File: helix-term/src/keymap.rs [hunk 17]
```diff
@@ -62,252 +283,298 @@ impl DerefMut for Keymaps {
...
+                "d" => surround_delete,
+                "a" => select_textobject_around,
+                "i" => select_textobject_inner,
+            },
+            "[" => { "Left bracket"
+                "d" => goto_prev_diag,
+                "D" => goto_first_diag,
+            },
+            "]" => { "Right bracket"
+                "d" => goto_next_diag,
```
Comment (sudormrfbin) [inline helix-term/src/keymap.rs:368]: `"Left bracket"` and `"Right bracket"` sound awkward...

Comment (archseer) [reply helix-term/src/keymap.rs:368]: Yeah we need a better name for these, they mimic unimpaired: https://github.com/tpope/vim-unimpaired

**Comment by archseer** [comment]
> BTreeMap sorts itself by keys, so it won't remember insertion order.

Ah right! I guess we'd need https://github.com/bluss/indexmap but I'm not sure if it's worth pulling in another dependency, a Vec of keys to sort by would work yeah.

File: helix-term/src/keymap.rs (hunk 14 above)
Comment (pickfire) [reply helix-term/src/keymap.rs]: Maybe I will take a look at this later myself.

Comment (pickfire) [reply helix-term/src/keymap.rs]: But can you please keep them as normal comments here?

File: helix-term/src/keymap.rs (hunk 17 above)
Comment (pickfire) [reply helix-term/src/keymap.rs:368]: I think the previous name for this is better.
```
┌select to next────────┐
│ d  diagnostic        │
│ D  diagnostic (last) │
└──────────────────────┘
```

Comment (archseer) [reply helix-term/src/keymap.rs:368]: It won't always select though (and in this case it jumps, not selects), it's just a bunch of pair mappings. `[<space>` could be `o<esc>` for example

File: helix-term/src/keymap.rs (hunk 14 above)
Comment (sudormrfbin) [reply helix-term/src/keymap.rs]: I changed the descriptions themselves in ad7e412

**Comment by sudormrfbin** [comment]
> > BTreeMap sorts itself by keys, so it won't remember insertion order.
> 
> Ah right! I guess we'd need [bluss/indexmap](https://github.com/bluss/indexmap) but I'm not sure if it's worth pulling in another dependency, a Vec of keys to sort by would work yeah.

Yeah pulling a dependency seems overkill, I went ahead and added it with a Vec :)

**Review by archseer — APPROVED** [review]
Great work!

**Comment by blaggacao** [comment]
@sudormrfbin Would it be possible to serialze the default keymap from a `runtime/default-keymap.toml`?

I need to rebind almost _all_ keys because I'm on a workman layout sometimes, so this is really annoying to do as a partial `rhs` merge since you always miss _one_ key somewhare that causes odd behavior.

With a `runtime/default-keymap.toml`, I could just copy that over and through editing discipline guarantee some decent consistency.

**Comment by cessen** [comment]
> I need to rebind almost all keys because I'm on a workman layout sometimes

Even aside from keyboard layouts, I think it would be good to support starting from a blank slate anyway, for building totally custom keymaps.
